freeipa/src/main/java/com/sequenceiq/freeipa/flow/instance/reboot/action/RebootActions.java

```diff
@@ -0,0 +1,149 @@
+import com.sequenceiq.cloudbreak.cloud.event.instance.RebootInstancesResult;
+import com.sequenceiq.cloudbreak.cloud.model.CloudCredential;
+import com.sequenceiq.cloudbreak.cloud.model.CloudInstance;
+import com.sequenceiq.cloudbreak.cloud.model.Location;
+import com.sequenceiq.cloudbreak.common.event.Selectable;
+import com.sequenceiq.cloudbreak.logger.MDCBuilder;
+import com.sequenceiq.flow.core.FlowParameters;
+import com.sequenceiq.freeipa.converter.cloud.CredentialToCloudCredentialConverter;
+import com.sequenceiq.freeipa.converter.cloud.InstanceMetaDataToCloudInstanceConverter;
+import com.sequenceiq.freeipa.dto.Credential;
+import com.sequenceiq.freeipa.entity.InstanceMetaData;
+import com.sequenceiq.freeipa.entity.Stack;
+import com.sequenceiq.freeipa.flow.instance.InstanceEvent;
+import com.sequenceiq.freeipa.flow.instance.InstanceFailureEvent;
+import com.sequenceiq.freeipa.flow.instance.reboot.RebootContext;
+import com.sequenceiq.freeipa.flow.instance.reboot.RebootEvent;
+import com.sequenceiq.freeipa.flow.instance.reboot.RebootService;
+import com.sequenceiq.freeipa.flow.instance.reboot.RebootState;
+import com.sequenceiq.freeipa.service.CredentialService;
+import com.sequenceiq.freeipa.service.stack.StackService;
+import com.sequenceiq.freeipa.service.stack.instance.InstanceMetaDataService;
+
+@Configuration
+public class RebootActions {
+    private static final Logger LOGGER = LoggerFactory.getLogger(RebootActions.class);
+
+    @Inject
+    private ConverterUtil converterUtil;
+
+    @Inject
+    private RebootService rebootService;
+
+    @Inject
+    private InstanceMetaDataToCloudInstanceConverter metadataConverter;
+
+    @Inject
+    private CredentialService credentialService;
+
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private InstanceMetaDataService instanceMetaDataService;
+
+    @Inject
+    private CredentialToCloudCredentialConverter credentialConverter;
+
+    @Bean(name = "REBOOT_STATE")
+    public Action<?, ?> rebootAction() {
+        return new AbstractRebootAction<>(InstanceEvent.class) {
+            @Override
+            protected void doExecute(RebootContext context, InstanceEvent payload, Map<Object, Object> variables) {
+                rebootService.startInstanceReboot(context);
+                sendEvent(context);
+                LOGGER.info("Starting reboot for {}", context.getInstanceMetaData().getInstanceId());
+            }
+
+            @Override
+            protected Selectable createRequest(RebootContext context) {
+                List<CloudInstance> cloudInstances = converterUtil.convertAll(List.of(context.getInstanceMetaData()), CloudInstance.class);
+                return new RebootInstancesRequest<>(context.getCloudContext(), context.getCloudCredential(), cloudInstances);
+            }
+
+            @Override
+            protected RebootContext createFlowContext(FlowParameters flowParameters, StateContext<RebootState, RebootEvent> stateContext,
```

Comment:
ok, so it's still a bit confusing here, so let me explain what I would do:
- InstanceEvent would contain stackId as resourceId as you need that to create the cloudcreadentials etc
- InstanceEvent would contain a set of instanceIds selected for reboot
- I would use `findAllInStack` for fetching instancemetadatas and filter in a stream, or have a separate repository query which contains stackId and instanceids too. Shouldn't make a big difference as freeipa won't have hundreds of instances
- You have to decide what happens if one/more/all instancemetadata is missing. I think this should not happen as it had been validated earlier, but maybe you should fail the flow with a message so it would be obvious that something is pretty wrong. Maybe you let in the force flag here and restart anyway the found instances and just log an error in that case
- if you really need these cloud stuff and real instancemetadata here then this override should be ok, so we wont make unnecessary db load for other steps.
- as this step is getting a big one, maybe you should move it a separate file
- if we send this event the restart will begin, but if we kill that service which is running this flow another will pick up from here and will send the reboot command again. Is it ok?

Reply:
refactored to have one event.  It is ok to reboot twice.